File: cpp/src/render/2d/choropleth_map/choropleth_map.cpp
/*
 * Copyright (C) 2019-2020 Zilliz. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <string>
#include <utility>
#include <vector>
#define GL_GLEXT_PROTOTYPES
#include <GL/gl.h>
#include <GL/glext.h>
#include <gdal_utils.h>
#include <ogrsf_frmts.h>

#include "render/2d/choropleth_map/choropleth_map.h"
#include "render/utils/color/color_gradient.h"

namespace arctern {
namespace render {

template class ChoroplethMap<int8_t>;

template class ChoroplethMap<int16_t>;

template class ChoroplethMap<int32_t>;

template class ChoroplethMap<int64_t>;

template class ChoroplethMap<uint8_t>;

template class ChoroplethMap<uint16_t>;

template class ChoroplethMap<uint32_t>;

template class ChoroplethMap<uint64_t>;

template class ChoroplethMap<float>;

template class ChoroplethMap<double>;

template <typename T>
ChoroplethMap<T>::ChoroplethMap(std::vector<OGRGeometry*> choropleth_wkb, T* count,
                                int64_t num_buildings)
    : choropleth_wkb_(std::move(choropleth_wkb)),
      count_(count),
      num_buildings_(num_buildings) {}

template <typename T>
void ChoroplethMap<T>::Draw() {
#ifndef USE_GPU
  glOrtho(0, window()->window_params().width(), 0, window()->window_params().height(), -1,
          1);
#endif

  glClear(GL_COLOR_BUFFER_BIT);

  glEnable(GL_BLEND);
  glBlendFunc(GL_SRC_ALPHA, GL_ZERO);

  for (int i = 0; i < num_buildings_; i++) {
    glColor4f(colors_[i * 4], colors_[i * 4 + 1], colors_[i * 4 + 2], colors_[i * 4 + 3]);
    glBegin(GL_POLYGON);
    for (int j = 0; j < buildings_x_[i].size(); j++) {
      glVertex2d(buildings_x_[i][j], buildings_y_[i][j]);
    }
    glEnd();
  }

  glFinish();
}

template <typename T>
void ChoroplethMap<T>::Transform() {
  buildings_x_.resize(num_buildings_);
  buildings_y_.resize(num_buildings_);

  for (int i = 0; i < num_buildings_; i++) {
    OGRGeometry* geometry = choropleth_wkb_[i];
    auto type = geometry->getGeometryType();
    if (type == OGRwkbGeometryType::wkbPolygon) {
      auto ring = geometry->toPolygon()->getExteriorRing();
      auto ring_size = ring->getNumPoints();
      buildings_x_[i].resize(ring_size);
      buildings_y_[i].resize(ring_size);
      for (int j = 0; j < ring_size; j++) {
        buildings_x_[i][j] = ring->getX(j);
        buildings_y_[i][j] = ring->getY(j);
      }
    } else {
      std::string err_msg = "Unknown geometry type";
      throw std::runtime_error(err_msg);
    }
    OGRGeometryFactory::destroyGeometry(geometry);
  }
}

template <typename T>
void ChoroplethMap<T>::SetColor() {
  colors_.resize(num_buildings_ * 4);

  auto count_start = choropleth_vega_.color_bound().first;
  auto count_end = choropleth_vega_.color_bound().second;
  auto count_range = count_end - count_start;

  size_t c_offset = 0;
  for (auto i = 0; i < num_buildings_; i++) {
    auto color_gradient = choropleth_vega_.color_gradient();
    if (color_gradient.size() == 1) {
      auto color = color_gradient[0];
      colors_[c_offset++] = color.r;
      colors_[c_offset++] = color.g;
      colors_[c_offset++] = color.b;
      colors_[c_offset++] = color.a;
    } else {
      auto count = count_[i] >= count_start ? count_[i] : count_start;
      count = count_[i] <= count_end ? count : count_end;
      auto ratio = (count - count_start) / count_range;
      auto color_start = color_gradient[0];
      auto color_end = color_gradient[1];
      auto color = ColorGradient::GetColor(color_start, color_end, ratio);
      colors_[c_offset++] = color.r;
      colors_[c_offset++] = color.g;
      colors_[c_offset++] = color.b;
      colors_[c_offset++] = color.a;
    }
  }
}

template <typename T>
std::vector<uint8_t> ChoroplethMap<T>::Render() {
  WindowsInit(choropleth_vega_.window_params());
  SetColor();
  Transform();
  Draw();
  Finalize();
  return Output();
}

}  // namespace render
}  // namespace arctern

File: python/zilliz_gis/util/vega/choropleth_map/choropleth_map.py
# Copyright (C) 2019-2020 Zilliz. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from zilliz_gis.util.vega.vega_node import (Width, Height, Description, Data,
                                            Scales, RootMarks, Root)


class Marks(RootMarks):
    """
        Top-Level Vega Specification Property: Marks
    """

    class Encode:
        class Value:
            def __init__(self, v: int or float or str):
                self.v = v

            def to_dict(self):
                dic = {
                    "value": self.v
                }
                return dic

        def __init__(self, bounding_box: Value, color_style: Value, ruler: Value, opacity: Value):
            if not (isinstance(bounding_box.v, list) and isinstance(color_style.v, str) and
                    isinstance(ruler.v, list) and isinstance(opacity.v, float)):
                # TODO error log here
                assert 0, "illegal"
            self._bounding_box = bounding_box
            self._color_style = color_style
            self._ruler = ruler
            self._opacity = opacity

        def to_dict(self):
            dic = {
                "enter": {
                    "bounding_box": self._bounding_box.to_dict(),
                    "color_style": self._color_style.to_dict(),
                    "ruler": self._ruler.to_dict(),
                    "opacity": self._opacity.to_dict()
                }
            }
            return dic

    def __init__(self, encode: Encode):
        self.encode = encode

    def to_dict(self):
        dic = [{
            "encode": self.encode.to_dict()
        }]
        return dic

class VegaChoroplethMap:
    def __init__(self, width: int, height: int,
                 bounding_box: list, color_style: str,
                 ruler: list, opacity: float):
        self._width = width
        self._height = height
        self._bounding_box = bounding_box
        self._color_style = color_style
        self._ruler = ruler
        self._opacity = opacity

    def build(self):
        description = Description(desc="building_weighted_2d")
        data = Data(name="data", url="/data/data.csv")
        domain = Scales.Scale.Domain("data", "c0")
        scale = Scales.Scale("building", "linear", domain)
        scales = Scales([scale])
        encode = Marks.Encode(bounding_box=Marks.Encode.Value(self._bounding_box),
                              color_style=Marks.Encode.Value(self._color_style),
                              ruler=Marks.Encode.Value(self._ruler),
                              opacity=Marks.Encode.Value(self._opacity))
        marks = Marks(encode)
        root = Root(Width(self._width), Height(self._height), description,
                    data, scales, marks)

        root_json = json.dumps(root.to_dict(), indent=2)
        return root_json
